invokeai/app/services/download/download_default.py

```diff
@@ -0,0 +1,367 @@
+
+import requests
+from pydantic.networks import AnyHttpUrl, Url
+from requests import HTTPError
+from tqdm import tqdm
+
+from invokeai.app.services.events.events_base import EventServiceBase
+from invokeai.app.util.misc import get_iso_timestamp
+from invokeai.backend.util.logging import InvokeAILogger
+
+from .download_base import (
+    DownloadEventHandler,
+    DownloadJob,
+    DownloadJobCancelledException,
+    DownloadJobStatus,
+    DownloadQueueServiceBase,
+    UnknownJobIDException,
+)
+
+# Maximum number of bytes to download during each call to requests.iter_content()
+DOWNLOAD_CHUNK_SIZE = 100000
+
+# marker that the queue is done and that thread should exit
+STOP_JOB = DownloadJob(id=-99, priority=-99, source=Url("http://dummy.com/"), dest=Path("/"))
+
+
+class DownloadQueueService(DownloadQueueServiceBase):
+    """Class for queued download of models."""
+
+    _jobs: Dict[int, DownloadJob]
+    _worker_pool: Set[threading.Thread]
+    _queue: PriorityQueue
+    _lock: threading.Lock
+    _logger: Logger
+    _events: Optional[EventServiceBase] = None
+    _next_job_id: int = 0
+    _requests: requests.sessions.Session
+
+    def __init__(
+        self,
+        max_parallel_dl: int = 5,
+        event_bus: Optional[EventServiceBase] = None,
+        requests_session: Optional[requests.sessions.Session] = None,
+    ):
+        """
+        Initialize DownloadQueue.
+
+        :param max_parallel_dl: Number of simultaneous downloads allowed [5].
+        :param requests_session: Optional requests.sessions.Session object, for unit tests.
+        """
+        self._jobs = {}
+        self._next_job_id = 0
+        self._queue = PriorityQueue()
+        self._worker_pool = set()
+        self._lock = threading.Lock()
+        self._logger = InvokeAILogger.get_logger("DownloadQueueService")
+        self._event_bus = event_bus
+        self._requests = requests_session or requests.Session()
+
+        self._start_workers(max_parallel_dl)
+
+    def __del__(self) -> None:
+        if any(thread.is_alive for thread in self._worker_pool):
+            self._queue.put(STOP_JOB)
+
+    def download(
+        self,
+        source: Union[str, AnyHttpUrl],
+        dest: Union[str, Path],
+        priority: int = 10,
+        access_token: Optional[str] = None,
+        on_start: Optional[DownloadEventHandler] = None,
+        on_progress: Optional[DownloadEventHandler] = None,
+        on_complete: Optional[DownloadEventHandler] = None,
+        on_error: Optional[DownloadEventHandler] = None,
+    ) -> DownloadJob:
+        """Create a download job and return its ID."""
+        with self._lock:
+            id = self._next_job_id
+            self._next_job_id += 1
+            job = DownloadJob(
+                id=id,
+                source=Url(source)
+                if isinstance(source, str)
+                else source,  # maybe a validator in DownloadJob would be better here
```

Comment:
Why not type `source` and `dest` args as `AnyHttpUrl` and `Path`, respectively? Seems reasonable to make the method require those types

Reply:
Fixed.